File: homework2/Saytam_Mishra/q14_floorInBST.java
package homework2.Saytam_Mishra;
// Technique - Search binary search tree (BST)
// Time Complexity - O(n) in worst case and O(logn) in average case
// Space Complexity - O(1)

public class q14_floorInBST {
    static class TreeNode {
        TreeNode left;
        TreeNode right;
        int val;

        public TreeNode(int val) {
            this.left = null;
            this.right = null;
            this.val = val;
        }
    }

    public static int floor(TreeNode root, int target) {
        TreeNode current = root;

        int ans = 0;
        boolean ifFound = false;
        while (current != null) {
            if (current.val <= target) {
                ifFound = true;
                ans = current.val;
                current = current.right;
            } else {
                current = current.left;
            }
        }

        return ifFound ? ans : -1;
    }

    public static void main(String[] args) {
        TreeNode root1 = new TreeNode(10);
        root1.left = new TreeNode(8);
        root1.right = new TreeNode(16);
        root1.left.right = new TreeNode(9);
        root1.right.left = new TreeNode(13);
        root1.right.right = new TreeNode(17);
        root1.right.right.right = new TreeNode(20);

        // Test Case 1

        int target1 = 13;
        int res1 = floor(root1, target1);
        System.out.println("Greatest element less than or equal to 13 is: " + (res1 != -1 ? res1 : "No element found"));

        // Test Case 2

        int target2 = 15;

        int res2 = floor(root1, target2);
        System.out.println("Greatest element less than or equal to 13 is: " + (res2 != -1 ? res2 : "No element found"));
    }
}

// Time spent - 9m 10s

File: homework2/Vedant_Pahariya/Q14_FloorInBST.cpp
// Given a target numeric value and a binary search tree, return the floor (greatest element less than or equal to the target) in the BST.

// Technique used: Search binary search tree recursively
// Time Complexity: O(h) where h is the height of the tree
// Space Complexity: O(h) due to recursion stack
// h = log n

#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;
};

int FloorinBST(Node* root, int target){
    if(root==NULL){
        return INT_MIN;
    }
    
    if(target > root->data){
        return max(root->data, FloorinBST(root->right, target));
    }
    if(target == root->data){
        return target;
    }
    if(target < root->data){
        return FloorinBST(root->left, target);
    }
}

int main(){
    Node* root = new Node();
    root->data = 10;
    root->left = new Node();
    root->left->data = 5;
    root->right = new Node();
    root->right->data = 15;
    root->left->left = new Node();
    root->left->left->data = 3;
    root->left->right = new Node();
    root->left->right->data = 7;
    root->right->right = new Node();
    root->right->right->data = 20;

    cout << FloorinBST(root, 12) << "\n"; // should return 10
    cout << FloorinBST(root, 5) << "\n";  // should return 5
    cout << FloorinBST(root, 2) << "\n";  // should return INT_MIN

    return 0;
}

// If a function returns INT_MIN, it means that there is no floor value in the BST for the given target. That is, all the values in the tree are greater than the target.